modules/ui.py

```diff
@@ -429,18 +430,132 @@ def refresh(request: gradio.routes.Request):
     return refresh_button
 
 
-def create_upload_button(label, elem_id, destination_dir):
-    def upload(file):
+def create_upload_button(label, elem_id, destination_dir, model_tracking_csv="models.csv"):
+
+    model_list_csv_path = os.path.join(destination_dir, model_tracking_csv)
+
+    def verify_model_existence(hash_str):
+        if os.path.exists(model_list_csv_path):
+            with open(model_list_csv_path) as csvfile:
+                modelreader = csv.reader(csvfile, delimiter=',')
+                for file_hash_str, file_name in modelreader:
+                    if hash_str == file_hash_str:
+                        return file_name
+                return None
+        return None
+
+    def upload_file(file, hash_str):
         file_path = file.name
-        shutil.move(file_path, destination_dir)
-        return
-    upload_button = gr.UploadButton(label=label, elem_id=elem_id, file_types=[".ckpt", ".safetensors", ".bin"])
+        new_path = shutil.move(file_path, destination_dir)
+        with open(new_path,"rb") as f:
+            file_bytes = f.read() # read entire file as bytes
+            readable_hash = hashlib.sha256(file_bytes).hexdigest()
+        if hash_str == readable_hash:
+            with open(model_list_csv_path, 'a') as csvfile:
+                modelwriter = csv.writer(csvfile, delimiter=',')
+                modelwriter.writerow([hash_str, new_path])
+            return new_path
+        return None
+    hash_str_id = elem_id+'-hash-str'
+    hash_str = gr.Textbox(label='hash str', elem_id=hash_str_id, visible=False)
+    existing_filepath = gr.Textbox(label='existing filepath', visible=False)
+    uploaded_filepath = gr.Textbox(label='uploaded filepath', visible=False)
+    button_id = "regular-button-" + elem_id
+    hidden_button_id = "hidden-button-" + elem_id
+    upload_button_id = "hidden-upload-button-" + elem_id
+    button = gr.Button(label, elem_id=button_id, variant="primary")
+
+    compute_hash_js = """
+        () => {{
+            const upload_button = document.querySelector(
+                "body > gradio-app").shadowRoot.querySelector(
+                "#{upload_button_id}");

```

Comment:
Just FYI: I saw a few "query selectors" used in this JS. I am not sure how fragile they are.